File: ncRepositorios/pParcela.cs
using System;
using System.Collections.Generic;
using Dapper;
using MySql.Data.MySqlClient;
using ncNComum;
using ncModelos;

namespace ncRepositorios
{
    public class pParcela : RepositorioBase, IpParcela
    {
        public ColecaoParcelas Listar()
        {
            try
            {
                using (var conn = CriarConexao())
                {
                    var lista = conn.Query<dParcelas>(
                        "SELECT cid, crediarioid AS crediarioId, codigoBarras, dataemissao AS dataEmissao, datavencimento AS dataVecimento, valor, valorreceber AS valorReceber, valorpago AS valorPago, situacao, observacao FROM Parcelas").AsList();
                    if (lista.Count == 0) return null;
                    var retorno = new ColecaoParcelas();
                    retorno.AddRange(lista);
                    return retorno;
                }
            }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex)
            {
                throw new ExcecaoNascomercio("Erro em Listar Crediário [" + ToString() + "] - " + ex.Message);
            }
        }

        public ColecaoParcelas Consultar(dParcelas dados)
        {
            try
            {
                using (var conn = CriarConexao())
                {
                    var conditions = new List<string>();
                    var p = new DynamicParameters();
                    if (dados.cid != 0) { conditions.Add("cid=@cid"); p.Add("cid", dados.cid); }
                    if (dados.crediarioId != 0) { conditions.Add("crediarioid=@crediarioId"); p.Add("crediarioId", dados.crediarioId); }
                    if (dados.codigoBarras != null && dados.codigoBarras.Trim() != "") { conditions.Add("codigoBarras=@codigoBarras"); p.Add("codigoBarras", dados.codigoBarras); }
                    var where = conditions.Count > 0 ? "WHERE " + string.Join(" AND ", conditions) : "";
                    var sql = $"SELECT cid, crediarioid AS crediarioId, codigoBarras, dataemissao AS dataEmissao, datavencimento AS dataVecimento, valor, valorreceber AS valorReceber, valorpago AS valorPago, situacao, datapagamento AS dataPagamento, observacao FROM parcelas {where}";
                    var lista = conn.Query<dParcelas>(sql, p).AsList();
                    if (lista.Count == 0) return null;
                    var retorno = new ColecaoParcelas();
                    retorno.AddRange(lista);
                    return retorno;
                }
            }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex)
            {
                throw new ExcecaoNascomercio("Erro em Consultar Crediario[" + ToString() + "] - " + ex.Message);
            }
        }

        public ColecaoParcelas ConsultarParcelasCliente(int codCliente)
        {
            try
            {
                using (var conn = CriarConexao())
                {
                    var conditions = new List<string>();
                    var p = new DynamicParameters();
                    if (codCliente != 0) { conditions.Add("clienteid=@clienteid"); p.Add("clienteid", codCliente); }
                    var where = conditions.Count > 0 ? "WHERE " + string.Join(" AND ", conditions) : "";
                    var sql = $"SELECT parcelas.cid, parcelas.crediarioid AS crediarioId, parcelas.codigoBarras, parcelas.dataemissao AS dataEmissao, parcelas.datavencimento AS dataVecimento, parcelas.valor, parcelas.valorreceber AS valorReceber, parcelas.valorpago AS valorPago, parcelas.situacao, parcelas.datapagamento AS dataPagamento, parcelas.observacao FROM parcelas INNER JOIN crediario ON parcelas.crediarioid = crediario.cid {where}";
                    var lista = conn.Query<dParcelas>(sql, p).AsList();
                    if (lista.Count == 0) return null;
                    var retorno = new ColecaoParcelas();
                    retorno.AddRange(lista);
                    return retorno;
                }
            }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex)
            {
                throw new ExcecaoNascomercio("Erro em Consultar Crediario[" + ToString() + "] - " + ex.Message);
            }
        }

        public ColecaoParcelas ConsultarParcelasVencidas(int codCliente)
        {
            try
            {
                using (var conn = CriarConexao())
                {
                    var conditions = new List<string>();
                    var p = new DynamicParameters();
                    if (codCliente != 0)
                    {
                        conditions.Add("clienteid=@clienteid");
                        p.Add("clienteid", codCliente);
                        conditions.Add("datavencimento < date(now())");
                    }
                    var where = conditions.Count > 0 ? "WHERE " + string.Join(" AND ", conditions) : "";
                    var sql = $"SELECT parcelas.cid, parcelas.crediarioid AS crediarioId, parcelas.codigoBarras, parcelas.dataemissao AS dataEmissao, parcelas.datavencimento AS dataVecimento, parcelas.valor, parcelas.valorreceber AS valorReceber, parcelas.valorpago AS valorPago, parcelas.situacao, parcelas.datapagamento AS dataPagamento FROM parcelas INNER JOIN crediario ON parcelas.crediarioid = crediario.cid {where} ORDER BY parcelas.datavencimento";
                    var lista = conn.Query<dParcelas>(sql, p).AsList();
                    if (lista.Count == 0) return null;
                    var retorno = new ColecaoParcelas();
                    retorno.AddRange(lista);
                    return retorno;
                }
            }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex)
            {
                throw new ExcecaoNascomercio("Erro em Consultar Crediario[" + ToString() + "] - " + ex.Message);
            }
        }

        public int Incluir(dParcelas dados)
        {
            try
            {
                using (var conn = CriarConexao())
                {
                    conn.Execute("INSERT INTO parcelas (crediarioid, codigoBarras, dataemissao, datavencimento, valor, valorreceber, valorpago, datapagamento, situacao, observacao) VALUES (@crediarioId, @codigoBarras, @dataEmissao, @dataVecimento, @valor, @valorReceber, @valorPago, @dataPagamento, @situacao, @observacao)",
                        new { dados.crediarioId, dados.codigoBarras, dados.dataEmissao, dados.dataVecimento, dados.valor, dados.valorReceber, dados.valorPago, dados.dataPagamento, dados.situacao, dados.observacao });
                    return (int)conn.ExecuteScalar<long>("SELECT LAST_INSERT_ID()");
                }
            }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex)
            {
                throw new ExcecaoNascomercio("Erro em Incluir Crediario [" + ToString() + "] - " + ex.Message);
            }
        }

        public ColecaoParcelas ConsultarPagamentos(dParcelas dados)
        {
            try
            {
                using (var conn = CriarConexao())
                {
                    var conditions = new List<string>();
                    var p = new DynamicParameters();
                    if (dados.crediarioId != 0) { conditions.Add("parcelasid=@parcelasid"); p.Add("parcelasid", dados.cid); }
                    var where = conditions.Count > 0 ? "WHERE " + string.Join(" AND ", conditions) : "";
                    var sql = $"SELECT parcelasid AS cid, valorpago AS valorPago, datapagamento AS dataPagamento, diasAtraso FROM parcelaspag {where}";
                    var lista = conn.Query<dParcelas>(sql, p).AsList();
                    if (lista.Count == 0) return null;
                    var retorno = new ColecaoParcelas();
                    retorno.AddRange(lista);
                    return retorno;
                }
            }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex)
            {
                throw new ExcecaoNascomercio("Erro em Consultar Pagamentos[" + ToString() + "] - " + ex.Message);
            }
        }

        public int IncluirPagamento(dParcelas dados)
        {
            try
            {
                using (var conn = CriarConexao())
                {
                    conn.Execute("INSERT INTO parcelaspag (parcelasid, valorpago, datapagamento, diasAtraso) VALUES (@cid, @valorPago, @dataPagamento, @diasAtraso)",
                        new { dados.cid, dados.valorPago, dados.dataPagamento, dados.diasAtraso });
                    return (int)conn.ExecuteScalar<long>("SELECT LAST_INSERT_ID()");
                }
            }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex)
            {
                throw new ExcecaoNascomercio("Erro em Incluir Pagamento [" + ToString() + "] - " + ex.Message);
            }
        }

        public int Corrigir()
        {
            try
            {
                using (var conn = CriarConexao())
                {
                    return conn.Execute("UPDATE parcelas SET valorreceber = 0.0 WHERE situacao = '01' AND valorreceber <= valorpago");
                }
            }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex)
            {
                throw new ExcecaoNascomercio("Erro em corrigir parcelas [" + ToString() + "] - " + ex.Message);
            }
        }

        public int Alterar(dParcelas dados)
        {
            try
            {
                using (var conn = CriarConexao())
                {
                    return conn.Execute("UPDATE parcelas SET codigoBarras=@codigoBarras, crediarioId=@crediarioId, dataEmissao=@dataEmissao, dataVencimento=@dataVecimento, valor=@valor, valorreceber=@valorReceber, valorpago=@valorPago, dataPagamento=@dataPagamento, situacao=@situacao, observacao=@observacao WHERE cid=@cid",
                        new { dados.codigoBarras, dados.crediarioId, dados.dataEmissao, dados.dataVecimento, dados.valor, dados.valorReceber, dados.valorPago, dados.dataPagamento, dados.situacao, dados.observacao, dados.cid });
                }
            }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex)
            {
                throw new ExcecaoNascomercio("Erro em Alterar crediario [" + ToString() + "] - " + ex.Message);
            }
        }

        public int Excluir(dParcelas dados)
        {
            try
            {
                using (var conn = CriarConexao())
                {
                    return conn.Execute("DELETE FROM parcelas WHERE cid=@cid", new { dados.cid });
                }
            }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex)
            {
                throw new ExcecaoNascomercio("Erro em Excluir crediario [" + ToString() + "] - " + ex.Message);
            }
        }
    }
}

File: ncPersistencia/pParcela.vb
Imports ncDados.nsCrediario
Imports ncComum.nsAcessoBD
Imports ncComum.nsFuncoes
Imports ncComum.nsExcecao

Namespace nsCrediario

    Public Class pParcela

        Public Function Listar() As ColecaoParcelas

            Dim retorno As ColecaoParcelas
            Dim acessoBanco As cAcessoBD
            Dim ds As DataSet
            Dim dt As DataTable
            Dim row As DataRow
            Dim item As dParcelas
            Dim comandoSQL As String


            Try

                acessoBanco = New cAcessoBD


                comandoSQL = " Select cid, crediarioid, codigoBarras, " & _
                             "dataemissao, datavencimento, valor, valorreceber, valorpago, situacao, observacao From Parcelas"

                ds = acessoBanco.ExecutarDS(comandoSQL)

                If Not ds Is Nothing Then
                    If ds.Tables.Count > 0 Then
                        dt = ds.Tables(0)

                        If dt.Rows.Count > 0 Then
                            retorno = New ColecaoParcelas

                            For Each row In dt.Rows
                                item = New dParcelas

                                item.cid = cFuncoes.RetornarInteiro(row("cid"))
                                item.codigoBarras = cFuncoes.RetornarTexto(row("codigoBarras"))
                                item.crediarioId = cFuncoes.RetornarInteiro(row("crediarioid"))
                                item.dataEmissao = cFuncoes.RetornarData(row("dataemissao"))
                                item.dataVecimento = cFuncoes.RetornarData(row("datavencimento"))
                                item.valor = cFuncoes.RetornarDecimal(row("valor"))
                                item.valorReceber = cFuncoes.RetornarDecimal(row("valorreceber"))
                                item.valorPago = cFuncoes.RetornarDecimal(row("valorpago"))
                                item.situacao = cFuncoes.RetornarTexto(row("situacao"))
                                item.observacao = cFuncoes.RetornarTexto(row("observacao"))

                                retorno.Add(item)
                            Next
                        Else
                            retorno = Nothing
                        End If
                    Else
                        retorno = Nothing
                    End If
                Else
                    retorno = Nothing
                End If

            Catch ex As Exception

                retorno = Nothing
                Throw New ExcecaoNascomercio("Erro em Listar Crediário [" & Me.ToString() & "] - " & ex.Message, ex)

            End Try

            Listar = retorno

        End Function

        Public Function Consultar(ByVal dados As dParcelas) As ColecaoParcelas

            Dim retorno As ColecaoParcelas
            Dim acessoBanco As cAcessoBD
            Dim ds As DataSet
            Dim dt As DataTable
            Dim row As DataRow
            Dim item As dParcelas
            Dim sqlSelect As String
            Dim sqlWhere As String
            Dim sqlFrom As String


            Try

                acessoBanco = New cAcessoBD


                sqlSelect = " Select cid, crediarioid, codigoBarras, " & _
                             "dataemissao, datavencimento, valor, valorreceber, valorpago, situacao, datapagamento, observacao "

                sqlWhere = String.Empty
                sqlFrom = " From parcelas "

                '-- data
                'If dados.cid <> 0 Then
                '  sqlWhere = cFuncoes.MontarParametrosSQL(sqlWhere, dados., "cid")
                'End If

                '-- cid
                If dados.cid <> 0 Then
                    sqlWhere = cFuncoes.MontarParametrosSQL(sqlWhere, dados.cid, "cid")
                End If

                '-- crediarioid
                If dados.crediarioId <> 0 Then
                    sqlWhere = cFuncoes.MontarParametrosSQL(sqlWhere, dados.crediarioId, "crediarioid")
                End If

                '-- codigoBarras
                If Not IsNothing(dados.codigoBarras) Then
                    If dados.codigoBarras.Trim <> "" Then
                        sqlWhere = cFuncoes.MontarParametrosSQL(sqlWhere, dados.codigoBarras, "codigoBarras")
                    End If
                End If

                If Not sqlWhere.Equals(String.Empty) Then
                    sqlWhere = " WHERE " & sqlWhere
                End If

                ds = acessoBanco.ExecutarDS(sqlSelect & " " & sqlFrom & " " & sqlWhere)

                If Not ds Is Nothing Then
                    If ds.Tables.Count > 0 Then
                        dt = ds.Tables(0)

                        If dt.Rows.Count > 0 Then
                            retorno = New ColecaoParcelas

                            For Each row In dt.Rows
                                item = New dParcelas

                                item.cid = cFuncoes.RetornarInteiro(row("cid"))
                                item.codigoBarras = cFuncoes.RetornarTexto(row("codigoBarras"))
                                item.crediarioId = cFuncoes.RetornarInteiro(row("crediarioid"))
                                item.dataEmissao = cFuncoes.RetornarData(row("dataemissao"))
                                item.dataVecimento = cFuncoes.RetornarData(row("datavencimento"))
                                item.valor = cFuncoes.RetornarDecimal(row("valor"))
                                item.valorReceber = cFuncoes.RetornarDecimal(row("valorreceber"))
                                item.valorPago = cFuncoes.RetornarDecimal(row("valorpago"))
                                item.dataPagamento = cFuncoes.RetornarData(row("datapagamento"))
                                item.situacao = cFuncoes.RetornarTexto(row("situacao"))
                                item.observacao = cFuncoes.RetornarTexto(row("observacao"))

                                retorno.Add(item)
                            Next
                        Else
                            retorno = Nothing
                        End If
                    Else
                        retorno = Nothing
                    End If
                Else
                    retorno = Nothing
                End If

            Catch ex As Exception

                retorno = Nothing
                Throw New ExcecaoNascomercio("Erro em Consultar Crediario[" & Me.ToString() & "] - " & ex.Message, ex)

            End Try

            Consultar = retorno

        End Function

        Public Function ConsultarParcelasCliente(ByVal codCliente As Integer) As ColecaoParcelas

            Dim retorno As ColecaoParcelas
            Dim acessoBanco As cAcessoBD
            Dim ds As DataSet
            Dim dt As DataTable
            Dim row As DataRow
            Dim item As dParcelas
            Dim sqlSelect As String
            Dim sqlWhere As String
            Dim sqlFrom As String

            Try

                acessoBanco = New cAcessoBD


                sqlSelect = " Select parcelas.cid, parcelas.crediarioid, parcelas.codigoBarras, " & _
                             "parcelas.dataemissao, parcelas.datavencimento, parcelas.valor," & _
                             "parcelas.valorreceber, parcelas.valorpago, parcelas.situacao, parcelas.datapagamento, parcelas.observacao "

                sqlWhere = String.Empty
                sqlFrom = " From parcelas inner join crediario on parcelas.crediarioid = crediario.cid"

                '-- codCliente
                If codCliente <> 0 Then
                    sqlWhere = cFuncoes.MontarParametrosSQL(sqlWhere, codCliente, "clienteid")
                End If

                If Not sqlWhere.Equals(String.Empty) Then
                    sqlWhere = " WHERE " & sqlWhere
                End If

                ds = acessoBanco.ExecutarDS(sqlSelect & " " & sqlFrom & " " & sqlWhere)

                If Not ds Is Nothing Then
                    If ds.Tables.Count > 0 Then
                        dt = ds.Tables(0)

                        If dt.Rows.Count > 0 Then
                            retorno = New ColecaoParcelas

                            For Each row In dt.Rows
                                item = New dParcelas

                                item.cid = cFuncoes.RetornarInteiro(row("cid"))
                                item.codigoBarras = cFuncoes.RetornarTexto(row("codigoBarras"))
                                item.crediarioId = cFuncoes.RetornarInteiro(row("crediarioid"))
                                item.dataEmissao = cFuncoes.RetornarData(row("dataemissao"))
                                item.dataVecimento = cFuncoes.RetornarData(row("datavencimento"))
                                item.valor = cFuncoes.RetornarDecimal(row("valor"))
                                item.valorReceber = cFuncoes.RetornarDecimal(row("valorreceber"))
                                item.valorPago = cFuncoes.RetornarDecimal(row("valorpago"))
                                item.dataPagamento = cFuncoes.RetornarData(row("datapagamento"))
                                item.situacao = cFuncoes.RetornarTexto(row("situacao"))
                                item.observacao = cFuncoes.RetornarTexto(row("observacao"))

                                retorno.Add(item)
                            Next
                        Else
                            retorno = Nothing
                        End If
                    Else
                        retorno = Nothing
                    End If
                Else
                    retorno = Nothing
                End If

            Catch ex As Exception

                retorno = Nothing
                Throw New ExcecaoNascomercio("Erro em Consultar Crediario[" & Me.ToString() & "] - " & ex.Message, ex)

            End Try

            Return retorno

        End Function

        Public Function ConsultarParcelasVencidas(ByVal codCliente As Integer) As ColecaoParcelas

            Dim retorno As ColecaoParcelas
            Dim acessoBanco As cAcessoBD
            Dim ds As DataSet
            Dim dt As DataTable
            Dim row As DataRow
            Dim item As dParcelas
            Dim sqlSelect As String
            Dim sqlWhere As String
            Dim sqlFrom As String

            Try

                acessoBanco = New cAcessoBD

                sqlSelect = " Select parcelas.cid, parcelas.crediarioid, parcelas.codigoBarras, " & _
                             "parcelas.dataemissao, parcelas.datavencimento, parcelas.valor," & _
                             "parcelas.valorreceber, parcelas.valorpago, parcelas.situacao, parcelas.datapagamento "

                sqlWhere = String.Empty
                sqlFrom = " From parcelas inner join crediario on parcelas.crediarioid = crediario.cid"

                '-- codCliente
                If codCliente <> 0 Then
                    sqlWhere = cFuncoes.MontarParametrosSQL(sqlWhere, codCliente, "clienteid")
                End If

                '-- data vencimento
                If codCliente <> 0 Then
                    sqlWhere = cFuncoes.MontarParametrosSQL(sqlWhere, " datavencimento < date(now()) ")
                End If

                If Not sqlWhere.Equals(String.Empty) Then
                    sqlWhere = " WHERE " & sqlWhere
                End If

                ds = acessoBanco.ExecutarDS(sqlSelect & " " & sqlFrom & " " & sqlWhere & " order by parcelas.datavencimento ")

                If Not ds Is Nothing Then
                    If ds.Tables.Count > 0 Then
                        dt = ds.Tables(0)

                        If dt.Rows.Count > 0 Then
                            retorno = New ColecaoParcelas

                            For Each row In dt.Rows
                                item = New dParcelas

                                item.cid = cFuncoes.RetornarInteiro(row("cid"))
                                item.codigoBarras = cFuncoes.RetornarTexto(row("codigoBarras"))
                                item.crediarioId = cFuncoes.RetornarInteiro(row("crediarioid"))
                                item.dataEmissao = cFuncoes.RetornarData(row("dataemissao"))
                                item.dataVecimento = cFuncoes.RetornarData(row("datavencimento"))
                                item.valor = cFuncoes.RetornarDecimal(row("valor"))
                                item.valorReceber = cFuncoes.RetornarDecimal(row("valorreceber"))
                                item.valorPago = cFuncoes.RetornarDecimal(row("valorpago"))
                                item.dataPagamento = cFuncoes.RetornarData(row("datapagamento"))
                                item.situacao = cFuncoes.RetornarTexto(row("situacao"))

                                retorno.Add(item)
                            Next
                        Else
                            retorno = Nothing
                        End If
                    Else
                        retorno = Nothing
                    End If
                Else
                    retorno = Nothing
                End If

            Catch ex As Exception

                retorno = Nothing
                Throw New ExcecaoNascomercio("Erro em Consultar Crediario[" & Me.ToString() & "] - " & ex.Message, ex)

            End Try

            Return retorno

        End Function

        Public Function Incluir(ByVal dados As dParcelas) As Integer

            Dim retorno As Integer
            Dim acessoBanco As cAcessoBD
            Dim comandoSQL As String


            Try

                acessoBanco = New cAcessoBD


                comandoSQL = " INSERT INTO " & _
                    " parcelas (crediarioid, codigoBarras, " & _
                             "dataemissao, datavencimento, valor, valorreceber, valorpago, datapagamento, situacao, observacao) " & _
                    " VALUES (" & _
                            cFuncoes.PersistirInteiro(dados.crediarioId) & "," & _
                            cFuncoes.PersistirTexto(dados.codigoBarras) & "," & _
                            cFuncoes.PersistirData(dados.dataEmissao) & "," & _
                            cFuncoes.PersistirData(dados.dataVecimento) & "," & _
                            cFuncoes.PersistirDecimal(dados.valor) & "," & _
                            cFuncoes.PersistirDecimal(dados.valorReceber) & "," & _
                            cFuncoes.PersistirDecimal(dados.valorPago) & "," & _
                            cFuncoes.PersistirData(dados.dataPagamento) & "," & _
                            cFuncoes.PersistirTexto(dados.situacao) & "," & _
                            cFuncoes.PersistirTexto(dados.observacao) & ")"

                retorno = acessoBanco.ExecutarCID(comandoSQL)

            Catch ex As Exception

                retorno = Nothing
                Throw New ExcecaoNascomercio("Erro em Incluir Crediario [" & Me.ToString() & "] - " & ex.Message, ex)

            End Try

            Return retorno

        End Function

        Public Function ConsultarPagamentos(ByVal dados As dParcelas) As ColecaoParcelas

            Dim retorno As ColecaoParcelas
            Dim acessoBanco As cAcessoBD
            Dim ds As DataSet
            Dim dt As DataTable
            Dim row As DataRow
            Dim item As dParcelas
            Dim sqlSelect As String
            Dim sqlWhere As String
            Dim sqlFrom As String


            Try

                acessoBanco = New cAcessoBD

                sqlSelect = " Select parcelasid, valorpago, datapagamento, diasAtraso " 

                sqlWhere = String.Empty
                sqlFrom = " From parcelaspag "

                '-- parcelasid
                If dados.crediarioId <> 0 Then
                    sqlWhere = cFuncoes.MontarParametrosSQL(sqlWhere, dados.cid, "parcelasid")
                End If

                If Not sqlWhere.Equals(String.Empty) Then
                    sqlWhere = " WHERE " & sqlWhere
                End If

                ds = acessoBanco.ExecutarDS(sqlSelect & " " & sqlFrom & " " & sqlWhere)

                If Not ds Is Nothing Then
                    If ds.Tables.Count > 0 Then
                        dt = ds.Tables(0)

                        If dt.Rows.Count > 0 Then
                            retorno = New ColecaoParcelas

                            For Each row In dt.Rows
                                item = New dParcelas

                                item.cid = cFuncoes.RetornarInteiro(row("parcelasid"))
                                item.valorPago = cFuncoes.RetornarDecimal(row("valorpago"))
                                item.dataPagamento = cFuncoes.RetornarData(row("datapagamento"))
                                item.diasAtraso = cFuncoes.RetornarInteiro(row("diasAtraso"))

                                retorno.Add(item)
                            Next
                        Else
                            retorno = Nothing
                        End If
                    Else
                        retorno = Nothing
                    End If
                Else
                    retorno = Nothing
                End If

            Catch ex As Exception

                retorno = Nothing
                Throw New ExcecaoNascomercio("Erro em Consultar Pagamentos[" & Me.ToString() & "] - " & ex.Message, ex)

            End Try

            Return retorno

        End Function

        Public Function IncluirPagamento(ByVal dados As dParcelas) As Integer

            Dim retorno As Integer
            Dim acessoBanco As cAcessoBD
            Dim comandoSQL As String


            Try

                acessoBanco = New cAcessoBD


                comandoSQL = " INSERT INTO " & _
                    " parcelaspag (parcelasid, valorpago, datapagamento, diasAtraso) " & _
                    " VALUES (" & _
                            cFuncoes.PersistirInteiro(dados.cid) & "," & _
                            cFuncoes.PersistirDecimal(dados.valorPago) & "," & _
                            cFuncoes.PersistirData(dados.dataPagamento) & "," & _
                            cFuncoes.PersistirInteiro(dados.diasAtraso) & ")"

                retorno = acessoBanco.ExecutarCID(comandoSQL)

            Catch ex As Exception

                retorno = Nothing
                Throw New ExcecaoNascomercio("Erro em Incluir Pagamento [" & Me.ToString() & "] - " & ex.Message, ex)

            End Try

            Return retorno

        End Function

        Public Function Corrigir() As Integer
            Dim retorno As Integer
            Dim acessoBanco As cAcessoBD
            Dim comandoSQL As String

            Try
                acessoBanco = New cAcessoBD

                comandoSQL = "update parcelas set valorreceber = 0.0 where situacao = '01' and valorreceber <= valorpago;"

                retorno = acessoBanco.ExecutarINT(comandoSQL)

            Catch ex As Exception

                retorno = Nothing
                Throw New ExcecaoNascomercio("Erro em corrigir parcelas [" & Me.ToString() & "] - " & ex.Message, ex)

            End Try

            Return retorno

        End Function

        Public Function Alterar(ByVal dados As dParcelas) As Integer

            Dim retorno As Integer
            Dim acessoBanco As cAcessoBD
            Dim comandoSQL As String

            Try

                acessoBanco = New cAcessoBD

                comandoSQL = " UPDATE parcelas SET " & _
                    " codigoBarras = " & cFuncoes.PersistirTexto(dados.codigoBarras) & "," & _
                    " crediarioId = " & cFuncoes.PersistirInteiro(dados.crediarioId) & "," & _
                    " dataEmissao = " & cFuncoes.PersistirData(dados.dataEmissao) & "," & _
                    " dataVencimento = " & cFuncoes.PersistirData(dados.dataVecimento) & "," & _
                    " valor = " & cFuncoes.PersistirDecimal(dados.valor) & "," & _
                    " valorreceber = " & cFuncoes.PersistirDecimal(dados.valorReceber) & "," & _
                    " valorpago = " & cFuncoes.PersistirDecimal(dados.valorPago) & "," & _
                    " dataPagamento = " & cFuncoes.PersistirData(dados.dataPagamento) & "," & _
                    " situacao = " & cFuncoes.PersistirTexto(dados.situacao) & "," & _
                    " observacao = " & cFuncoes.PersistirTexto(dados.observacao) & _
                    " WHERE " & _
                    " cid = " & dados.cid.ToString()

                retorno = acessoBanco.ExecutarINT(comandoSQL)

            Catch ex As Exception

                retorno = Nothing
                Throw New ExcecaoNascomercio("Erro em Alterar crediario [" & Me.ToString() & "] - " & ex.Message, ex)

            End Try

            Return retorno

        End Function

        Public Function Excluir(ByVal dados As dParcelas) As Integer

            Dim retorno As Integer
            Dim acessoBanco As cAcessoBD
            Dim comandoSQL As String

            Try

                acessoBanco = New cAcessoBD

                comandoSQL = " DELETE FROM parcelas " & _
                    " WHERE cid = " & dados.cid.ToString()

                retorno = acessoBanco.ExecutarINT(comandoSQL)

            Catch ex As Exception

                retorno = Nothing
                Throw New ExcecaoNascomercio("Erro em Excluir crediario [" & Me.ToString() & "] - " & ex.Message, ex)

            End Try

            Excluir = retorno

        End Function

    End Class

End Namespace
